app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/PeerGroupHandshakeAsyncTask.java

```diff
@@ -0,0 +1,138 @@
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver. Finally, the onPostExecute() of the sender
+ * initiates the file transfer through {@link SenderDeviceAsyncTask} on the sender and using
+ * {@link ReceiverDeviceAsyncTask} on the receiver.
+ * */
+class PeerGroupHandshakeAsyncTask extends AsyncTask<Void, Void, InetAddress> {
+
+  private static final String TAG = "PeerGrpHndshakeAsyncTsk";
+  private final String HANDSHAKE_MESSAGE = "Request Kiwix File Sharing";
+
+  private DeviceListFragment deviceListFragment;
+  private WifiP2pInfo groupInfo;
+
+  public PeerGroupHandshakeAsyncTask(DeviceListFragment deviceListFragment, WifiP2pInfo groupInfo) {
+    this.deviceListFragment = deviceListFragment;
+    this.groupInfo = groupInfo;
+  }
+
+  @Override
+  protected InetAddress doInBackground(Void... voids) {
+
+    try {
+      if(groupInfo.groupFormed && groupInfo.isGroupOwner) {
+        ServerSocket serverSocket = new ServerSocket(PEER_HANDSHAKE_PORT);
+        serverSocket.setReuseAddress(true);
+        Socket server = serverSocket.accept();
+
+        ObjectInputStream objectInputStream = new ObjectInputStream(server.getInputStream());
+        Object object = objectInputStream.readObject();
+
+        // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+        if (object.getClass().equals(String.class) && object.equals(HANDSHAKE_MESSAGE)) {
+          if(BuildConfig.DEBUG) Log.d(TAG, "Client IP address: "+ server.getInetAddress());
+
+          exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream());
+
+          server.close();
+
+          return server.getInetAddress();
+        }
+      }
+      else if(groupInfo.groupFormed) { //&& !groupInfo.isGroupOwner
+        Socket client = new Socket();
+        client.setReuseAddress(true);
+        client.connect((new InetSocketAddress(groupInfo.groupOwnerAddress.getHostAddress(), PEER_HANDSHAKE_PORT)), 15000);
+
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(client.getOutputStream());
+        objectOutputStream.writeObject(HANDSHAKE_MESSAGE); // Send message for the peer device to verify
+
+        exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream());
+
+        client.close();
+
+        return groupInfo.groupOwnerAddress;
+      }
+
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    return null; // So a null is only returned in case of an exception
+  }
+
+  private void exchangeFileTransferMetadata(OutputStream outputStream, InputStream inputStream) {
+    try {
+      if(deviceListFragment.isFileSender()) {
+        ObjectOutputStream objectOutputStream =  new ObjectOutputStream(outputStream);
```

Comment:
This doesn't get closed? Use `try-with-resources`

Reply:
`try-with-resources` was news for me, thanks!
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/ab9f9368f35c33ddce721678ac2326e2d753ae62#diff-b6bbaac42877486339bfdb7cc97c9aa4R105)